database-commons/src/main/java/co/cask/db/batch/sink/AbstractDBSink.java

```diff
@@ -221,18 +218,60 @@ private boolean tableExists(Class<? extends Driver> jdbcDriverClass, String tabl
       throw Throwables.propagate(e);
     }
 
-    try (Connection connection = DriverManager.getConnection(connectionString, dbSinkConfig.getConnectionArguments())) {
-      DatabaseMetaData metadata = connection.getMetaData();
-      try (ResultSet rs = metadata.getTables(null, null, tableName, null)) {
-        return rs.next();
+    try (Connection connection = DriverManager.getConnection(connectionString, dbSinkConfig.getConnectionArguments());
+         ResultSet tables = connection.getMetaData().getTables(null, null, tableName, null)) {
+
+      Preconditions.checkArgument(tables.next(), "Table %s does not exist. " +
+                                  "Please check that the 'tableName' property " +
+                                  "has been set correctly, and that the connection string %s " +
+                                  "points to a valid database.",
+                                  tableName, connectionString);
+
+      try (PreparedStatement pStmt = connection.prepareStatement("SELECT * FROM " + tableName + " WHERE 1 = 0");
+           ResultSet rs = pStmt.executeQuery()) {
+        validateFields(inputSchema, rs);
       }
+
     } catch (SQLException e) {
-      LOG.error("Exception while trying to check the existence of database table {} for connection {}.",
+      LOG.error("Exception while trying to validate schema of database table {} for connection {}.",
                 tableName, connectionString, e);
       throw Throwables.propagate(e);
     }
   }
 
+  private void validateFields(Schema inputSchema, ResultSet rs) throws SQLException {
+    ResultSetMetaData rsMetaData = rs.getMetaData();
+
+    Preconditions.checkNotNull(inputSchema.getFields());
+    List<String> invalidFields = new ArrayList<>();
+
+    for (Schema.Field field : inputSchema.getFields()) {
+      int columnIndex = rs.findColumn(field.getName());
+      int type = rsMetaData.getColumnType(columnIndex);
+      int precision = rsMetaData.getPrecision(columnIndex);
+      int scale = rsMetaData.getScale(columnIndex);
+
+      Schema columnSchema;
+      if (ResultSetMetaData.columnNullable == rsMetaData.isNullable(columnIndex)) {
+        columnSchema = Schema.nullableOf(DBUtils.getSchema(type, precision, scale));
+      } else {
+        columnSchema = DBUtils.getSchema(type, precision, scale);
+      }
+
+      // we check for null-assignment compatibility, because schema.isCompatible() doesn't
+      // handle this case correctly
+      boolean isNotNullAssignable = !columnSchema.isNullable() && field.getSchema().isNullable();
+      boolean isNotCompatible = !field.getSchema().isCompatible(columnSchema);
+
+      if (isNotCompatible || isNotNullAssignable) {
+        invalidFields.add(field.getName());
+      }
+    }
+
+    Preconditions.checkArgument(invalidFields.isEmpty(), "Couldn't find matching database column(s) " +
```

Comment:
This error message won't give the full information about what went wrong here (does the column not exist, is the field nullable while the column is not nullable, do the types not match, etc). 

It's better to error log the problems as we come to them in addition to throwing this exception.

Reply:
OK